specs/agents/tracing-instrumentation-azure-functions.md

```diff
@@ -0,0 +1,109 @@
+
+### Disabled Functionality
+
+Agents SHOULD detect whether they are running in an Azure Functions environment by testing
+for the existence of the `FUNCTIONS_WORKER_RUNTIME` environment variable.
+
+The following agent functionalities SHOULD to be turned off when tracing Azure Functions until decided otherwise:
+
+* **Metrics collection:** this includes all kind of metrics: system, process and breakdown metrics and is equivalent to
+setting `ELASTIC_APM_METRICS_INTERVAL = 0`
+* **Remote configuration:** equivalent to setting `ELASTIC_APM_CENTRAL_CONFIG = false`
+* **Cloud metadata discovery:**
+  * Equivalent to setting `ELASTIC_APM_CLOUD_PROVIDER = none`
+  * Instead an Azure Functions specific metadata detection MUST be performed (see see [Metadata](./metadata.md)).
+* **System metadata discovery:** in some agents, this may be a relatively heavy task. For example, the Java agent
+executes external commands in order to discover the hostname, which is not required for tracing Azure Functions. All other
+agents read and parse files to extract container and k8s metadata, which is not required as well.
+
+## Trigger-Specific Instrumentation
+
+Azure Functions can be triggered in [many different ways](https://learn.microsoft.com/en-us/azure/azure-functions/functions-triggers-bindings).
+A generic transaction for an Azure Functions invocation can be created independently of the actual trigger
+based on the [generic instrumentation](#generic-instrumentation) fields described above.
+
+However, depending on the trigger type, different information might be available that can be used
+to capture additional transaction data or that allows additional, valuable spans to be derived.
+
+### HTTP Trigger
+
+HTTP invocations typically provide *some* variants of an HTTP *request object* (e.g. [`HttpRequestData`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.identitymodel.protocols.httprequestdata))
+and an HTTP *response object* (e.g. [`HttpResponseData`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.azure.functions.worker.http.httpresponsedata)).
+
+Agents SHOULD use the information in the *request* and *response objects* to
+fill the HTTP context (`context.request` and `context.response`) fields in the same way it is done for HTTP transactions.
+
+In particular, agents MUST use HTTP headers to retrieve the `traceparent` and the `tracestate`
+and use those to start the transaction for the tracing the function execution.
+
+In addition the following fields should be set for HTTP trigger invocations:
+
+| Field | Value | Description | Source |
+| - | - | - | - |
+| `type` | `request`| Transaction type. Constant value for HTTP trigger invocations. | |
+| `name` | e.g. `GET products/{category:alpha}/{id:int?}`, `GET /api/SampleHttpTrigger` | Transaction name: HTTP method followed by a whitespace and low cardinality representation of the function (route pattern or resource path). | *request object* |
```

Comment:
Do we have any plans/ideas for how to get the low-cardinality name? `HttpRequestData` appears to only have the request uri, not any route information.

Reply:
Pragmatically, I would say this is best-effort depending on what's provided for the language/runtime in question.
As @trentm [pointed out](https://github.com/elastic/apm/pull/716#discussion_r1023405630), this information seems to be available in Node.js.

Reply:
> are you able to get that route from within the runtime?

@basepi  I'm not sure that it is. My earlier comment was an observation of the config file from the developer side. I haven't yet played with real deployed Azure Fns.

Reply:
(/cc @beniwohli)

I have some alternative suggestions for `transaction.name` for discussion. Sorry this got long.

An Azure Function app deployment looks like this (at least a Node.js one on windows
using [`WEBSITE_RUN_FROM_PACKAGE=1`](https://learn.microsoft.com/en-us/azure/azure-functions/run-functions-from-deployment-package)):

    C:\home\site\wwwroot\
        host.json           // Config for the function app
        HttpFn1\          // This dirname is the "function name"
            function.json   // Some config for the function, including a "route" template
            index.js        // The code for the function.
        HttpFn2\
            function.json   // Say has `"route": "products/{category:alpha}/{id:int?}"`
            index.js

Obviously "index.js" differs for other languages. My understanding is that "function.json" is a generated file for the compiled languages.

For HTTP-triggered functions, the URL to invoke it is:

    http(s)://<HOST:PORT><PREFIX>/<ROUTE>

The default PREFIX is `/api` and the default ROUTE is the function name.
For example:

    http://localhost:7071/api/HttpFn1     # locally using `func start`
    https://trentm-play-az-fn-app-1.azurewebsites.net/api/httpfn1

Notes ([Azure docs link](https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-http-webhook-trigger?tabs=in-process%2Cfunctionsv2&pivots=programming-language-javascript#customize-the-http-endpoint)):

- The PREFIX can be configured (including to the empty string) via `extensions.http.routePrefix` in "host.json".
- The ROUTE can be configured in "function.json" (or in some code annotations, presumably).
- Also, note that the path (at least for *Windows*-deployed apps) is case-insensitive (yeah IIS), so we should avoid using the request URL path for the transaction name.


The currently spec'd examples:

> `GET products/{category:alpha}/{id:int?}`, `GET /api/SampleHttpTrigger`

are inconsistent. One is `<METHOD> <ROUTE>`, one is `<METHOD> <PREFIX>/<ROUTE>`.
I think we should consider one of the following.

## option A: `METHOD ROUTE`

Examples:

    GET HttpFn1
    GET products/{category:alpha}/{id:int?}

This would require instrumentation code to get the possible route template. The only way I found to get this was to read and parse the "function.json" on first invocation of a particular function, and then cache that. My [implementation is here](https://github.com/elastic/apm-agent-nodejs/blob/c0c2d0995e4edd6cd868872064ee9ccc29f9506e/lib/instrumentation/modules/azure-functions-nodejs-worker/dist/src/worker-bundle.js#L70-L102).

This code relies on the `context.executionContext.functionDirectory` field, which is perhaps notably *not* defined on the Java "ExecutionContext" interface: https://learn.microsoft.com/en-us/java/api/com.microsoft.azure.functions.executioncontext?view=azure-java-stable
*Perhaps* the path `C:\home\site\wwwroot\${executionContext.functionName}\function.json` could be assumed?

<img width="391" alt="Screen Shot 2022-12-19 at 5 00 22 PM" src="https://user-images.githubusercontent.com/46866/208556903-94349136-2105-40cc-a492-6b7bc6473c4e.png">



## option B: `METHOD PREFIX/ROUTE`

Examples:

    GET /api/HttpFn1
    GET /api/products/{category:alpha}/{id:int?}

This would require instrumentation code to also load the "host.json" file to get a possibly customized `routePrefix`. I haven't implemented this yet.

---

Thoughts?

Reply:
Right, I think having the full route is preferable here.
Especially in cases where developers change the default from `api` to something else, they would expect this name to be reflected in the span names.

(As for the mentioned inconsistency - this was not my intention. I meant for the full route to be present in both cases in the beginning).